File: 2018/12-Dec/12-22-multiple-inheritance/swift/trees.swift
#!/usr/bin/swift

/*
 Experiment with multiple inheritance (with protocols) in Swift
 - Author: Andrew Jarombek
 - Date: 12/16/2018
 */

import Foundation

/**
 Create a protocol for a generic tree.  Protocols are similar to interfaces from Java
 */
protocol Tree {

    /**
     Get the height of the tree
     */
    func height() -> [String:Int]
}

typealias ChristmasTree = Tree

/**
 Add an extension to the Tree protocol for Christmas trees
 */
extension ChristmasTree {

    var type: String {
        return "Christmas"
    }
}

typealias EvergreenTree = Tree

/**
 Add an extension to the Tree protocol for Evergreen trees
 */
extension EvergreenTree {
    // Can't redeclare var 'type', since it already exists in ChristmasTree

    var leaf_persistence: String {
        return "Evergreen"
    }
}

/**
 Implement the Tree protocol with the ChristmasTree and EvergreenTree extensions
 */
class BalsamFir: Tree {

    let feet: Int
    let inches: Int

    init(height: [String:Int]) {
        self.feet = height["feet"] ?? 0
        self.inches = height["inches"] ?? 0
    }

    /**
     Implement the height() method defined in the Tree protocol.
     */
    func height() -> [String:Int] {
        return ["feet": self.feet, "inches": self.inches]
    }

    /**
     Determine whether or not the leaves on the tree persist through winter
     */
    func leafPersistence() -> Bool {
        return self.leaf_persistence == "Evergreen"
    }
}

// Test scripts

let balsam = BalsamFir(height: ["feet": 7, "inches": 2])

assert(balsam.type == "Christmas")
assert(balsam.height() == ["feet": 7, "inches": 2])
assert(balsam.leafPersistence())

File: 2018/12-Dec/12-22-multiple-inheritance/cpp/trees.cpp
/**
 * Explore multiple inheritance in C++
 * @author Andrew Jarombek
 * @date 12/17/2018
 */

#include <iostream>

using namespace std;

// Defines a generic tree
class Tree {
protected:
    int feet;
    int inches;
public:
    Tree(int feet, int inches) {
        cout << "Tree()" << endl;
        this->feet = feet;
        this->inches = inches;
    }

    // Calculates the height of the tree in inches
    int height() {
        return (this->feet * 12) + this->inches;
    }
};

// Defines a Christmas tree
class ChristmasTree: virtual public Tree {
public:
    ChristmasTree(int feet, int inches): Tree(feet, inches) {
        cout << "ChristmasTree()" << endl;
    }

    string type() {
        return "Christmas";
    }
};

// Defines an evergreen tree
class EvergreenTree: virtual public Tree {
public:
    EvergreenTree(int feet, int inches): Tree(feet, inches) {
        cout << "EvergreenTree()" << endl;
    }

    string type() {
        return "Evergreen";
    }
};

// Defines a balsam fir, which is a christmas tree and an evergreen tree
class BalsamFir: public ChristmasTree, public EvergreenTree {
public:
    BalsamFir(int feet, int inches): ChristmasTree(feet, inches), EvergreenTree(feet, inches), Tree(feet, inches) {
        cout << "BalsamFir()" << endl;
    }

    bool leafPeristence() {
        return EvergreenTree::type() == "Evergreen";
    }
};

int main() {
    BalsamFir balsamFir(7, 2);

    assert(balsamFir.height() == 86);
    assert(balsamFir.leafPeristence());

    // Calling type() is ambiguous since both ChristmasTree and EvergreenTree have
    // assert(balsamFir.type());

    return 0;
}